src/basicfuns.jl

```diff
@@ -152,9 +152,38 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x ≤ x0
+        return exp(x)
+    elseif x ≤ x1
+        return log1p(exp(x))
+    elseif x ≤ x2
+        return x + exp(-x)
+    else
+        return oftype(x0, x)
```

Comment:
Not needed with the suggestion above:
```suggestion
        return x
```